feat: enable public output note creation during NTX execution

partylikeits1983 opened the pull request:
### Brief Summary: This PR adds the required functionality to enable the creation of public output notes during network transactions. 

TODO list:
- [x] Update `build_recipient` procedure to push `SCRIPT_ROOT` into `AdviceMap`
- [x] Modified `on_note_after_created()` in `TransactionBaseHost` - Now checks if note script exists in `AdviceMap`
- [x] Added `.add_note_script()` method to `TransactionContextBuilder` - mockchain method for testing out adding note script to `TransactionExecutorBase`
- [x] Added `on_note_script_requested()` handler in TransactionExecutorHost: Fetches missing note scripts from `DataStore` and adds them to `AdviceMap`

Related to #1748
Resolves #1972

After this PR is merged, the `MINT` note script in #1925 will need to be updated so that its `NoteInputs` contain the note metadata of the public output note to create during the network transaction. Will open a separate issue to address this once this is merged or close to being merged.

partylikeits1983 commented on crates/miden-testing/src/tx_context/builder.rs:
```diff
@@ -303,6 +303,12 @@ impl TransactionContextBuilder {
+        for (script_root, note_script) in &self.note_scripts {
+            let script_felts: Vec<miden_processor::Felt> = note_script.into();
+            tx_args.extend_advice_map([(*script_root, script_felts)]);
+        }
```
I didn't initially realize that I wasn't using the lazy loading pattern. Removed this snippet, now lazy loading of the note script is working.

partylikeits1983 commented on crates/miden-tx/src/host/mod.rs:
```diff
@@ -510,22 +533,61 @@ where
+                Word::new([data[5], data[6], data[7], data[8]])
+            } else if data.len() == 12 {
+                // New format: [SERIAL_NUM, SCRIPT_ROOT, INPUTS_HASH]
+                Word::new([data[4], data[5], data[6], data[7]])
```
Cleaned up this logic, now there's only 1 format

partylikeits1983 commented on crates/miden-tx/src/host/mod.rs:
```diff
@@ -510,22 +533,61 @@ where
         process: &ProcessState,
-    ) -> Result<(), TransactionKernelError> {
+    ) -> Result<TransactionEventHandling, TransactionKernelError> {
         let stack = process.get_stack_state().split_off(1);
```
Applied these suggestions.

partylikeits1983 commented on crates/miden-lib/asm/miden/note.masm:
```diff
@@ -108,14 +108,56 @@ end
+
+    # inserting NoteDetails into AdviceMap
+    mem_storew.0 dropw mem_storew.4 dropw mem_storew.8 dropw
+    # => [RECIPIENT]
```
I think I addressed all of your comments here, except for removing `inputs_length` from the `RECIPIENT` advice entry. I think keeping `inputs_length` in the `RECIPIENT` entry keeps things on the Rust side in the host slightly simpler. What do you think?

PhilippGackstatter replied:
> I think keeping `inputs_length` in the `RECIPIENT` entry keeps things on the Rust side in the host slightly simpler. What do you think?

Yeah that's totally fine. I'm not sure if we really need it or have it in other places for advice map entries, so being consistent would be nice, but this is something we can address separately if at all.

PhilippGackstatter commented on crates/miden-tx/src/host/mod.rs:
```diff
@@ -219,6 +217,31 @@ where
+        _script_root: Word,
+        _script_felts: Vec<Felt>,
+        adv_provider: &miden_processor::AdviceProvider,
+    ) -> Result<(), TransactionKernelError> {
```
I hadn't realized before that we need to mutate the advice provider to be able to use the `OutputNoteBuilder`.

With the changes here, I think a slightly better structure would be to change `OutputNoteBuilder::new` to:

```rust
pub fn new(
    metadata: NoteMetadata,
    recipient_digest: Word,
    script: Option<NoteScript>,
    adv_provider: &AdviceProvider,
) -> Result<Self, TransactionKernelError> { ... }
```

And then:
- In `on_note_after_created`, we call this with `script: None`.
- In `on_note_script_requested` we already have the `NoteScript` that we got from the data store and we can pass it to the output note builder directly. Then if `script.is_some()` it can just use that directly and does not need to `NoteScript::try_from(script_data)`.

That way, we don't need to clone the entire advice provider, which is the main benefit.

This still feels a bit suboptimal. It would be nice if we could do all advice-related operations outside of the output note builder, but that would just shift the recipient extraction to somewhere else and I can't come up with a nice structure for this, so I would suggest the above.

PhilippGackstatter replied:
Actually, I think passing the `process` inside the `async move` block in the executor host shouldn't work for the no-std target, not sure why it does. I think we should avoid it regardless. This means we cannot use the advice provider there and we need the output note builder to work without it.

I would overall then change it like this. `OutputNoteBuilder::new`:

```rust
pub fn new(
    metadata: NoteMetadata,
    recipient: Option<NoteRecipient>,
) -> Result<Self, TransactionKernelError> {
```

Then, in `on_note_after_created`:
- Check if the advice provider contains the `recipient_data`.
- If not:
    - And `!metadata.is_private()`, return an error.
    - And `metadata.is_private()` build an output note builder with `recipient = None`. 
- If the data is present, extract the advice recipient data into:
    - serial number (`Word`)
    - note metadata (`NoteMetadata`)
    - script_root (`Word`) and maybe note script (`Option<NoteScript>`)
- If the note script is present, build an output note builder with the constructed `NoteRecipient`.
- If the note script is not present, return
    ```rust
    TransactionEventHandling::Unhandled(TransactionEventData::NoteScript {
        note_metadata
        serial_number,
        script_root,
        // ... (possibly other needed data)
    })
    ```

Then in `on_note_script_requested`:
- Request the note script from the data store.
- If the data store does not have the script and `metadata.is_private()`, do not error but build an output note builder with `recipient = None`. Otherwise, return an error.
- If the data store returns the `NoteScript`, build the `NoteRecipient` and call `OutputNoteBuilder::new`.

This may be missing some details, but I hope the overall structure becomes reasonably clear. 

This is mainly the same logic as before, just differently organized. In particular, we should handle the case where the data store does not have the script but the note is private gracefully and not error. I.e. we have a best effort approach for private notes.

partylikeits1983 commented on crates/miden-tx/src/host/mod.rs:
```diff
@@ -509,23 +538,79 @@ where
+                self.output_notes.len(),
+                note_idx
+            )));
+        }
```
Removed the duplicate check

mmagician commented on crates/miden-lib/asm/miden/note.masm:
```diff
@@ -101,21 +106,76 @@ end
+    loc_storew.12 dropw loc_storew.8 dropw loc_storew.4 dropw
+    # => [RECIPIENT]
+
+    loc_load.1 loc_store.3
```
what is the function of this line?

partylikeits1983 replied:
Copying memory in local store slot 1 to slot 3. Simplified to initially write to slot 3.

mmagician commented on crates/miden-testing/tests/scripts/faucet.rs:
```diff
@@ -275,3 +276,163 @@ async fn prove_burning_fungible_asset_on_existing_faucet_succeeds() -> anyhow::R
+/// The test creates a note that calls the faucet's `distribute` function to create a PUBLIC
+/// P2ID output note. The P2ID script is fetched from the data store during transaction execution.
+#[tokio::test]
+async fn test_public_note_creation_with_script_from_datastore() -> anyhow::Result<()> {
```
AFAICS this test doesn't cover the functionality proposed in the PR, namely of **public output note creation during NTX execution**. The account against which we execute the transactions is a `Public`, not `Network` account.

(Actually this test, when ported to `next`, works with a small change)
```rust
let executed_transaction = mock_chain
    .build_tx_context(faucet.id(), &[trigger_note.id()], &[])?
    // .add_note_script(p2id_script)
    .extend_expected_output_notes(vec![OutputNote::Full(expected_p2id_note)])
    .build()?
    .execute()
    .await?;

```

partylikeits1983 replied:
Technically you are correct since the faucet creating the note is a basic faucet with public storage. This test is set up similarly to how the `MINT` note works on the ntx faucet PR which is still waiting to be merged. 

A network transaction is the same as a regular tx, its just the network transaction builder executes an arbitrary public note against a public account. This test is testing that a public note can be created from a transaction that does not specify the public output notes in Rust, but instead supplies the `SCRIPT_ROOT` on the masm side. The `NOTE_SCRIPT` is in the `DataStore` under the `SCRIPT_ROOT` as the key.

mmagician replied:
All right, I see.
And for ntx transactions we can't specify the public output notes in Rust?

PhilippGackstatter commented on crates/miden-testing/tests/scripts/faucet.rs:
```diff
@@ -275,3 +289,153 @@ async fn prove_burning_fungible_asset_on_existing_faucet_succeeds() -> anyhow::R
+                # => [amount, tag, aux, note_type, execution_hint, RECIPIENT]
+
+                call.::miden::contracts::faucets::basic_fungible::distribute
+                # => [note_idx, pad(15)]
```
Is it essential for the test to call `distribute` instead of `output_note::create`?

bobbinth commented on crates/miden-tx/src/executor/exec_host.rs:
```diff
@@ -362,6 +364,106 @@ where
+
+                let inputs = match inputs_data {
+                    None => NoteInputs::default(),
+                    Some(inputs) => {
```
Similar code is used in another place as well. Could we create a helper function that encapsulates this logic? It could look something like:
```rust
fn read_recipient_info(
    advice: &AdviceProvider,
    recipient_digest: Word,
) -> Result<(NoteInputs, Word, Word), TransactionKernelError> {
    // return (NoteInputs, script_root, serial_num)
}
```

bobbinth commented on crates/miden-tx/src/executor/exec_host.rs:
```diff
@@ -362,6 +364,102 @@ where
+                let inputs = match inputs_data {
+                    None => NoteInputs::default(),
+                    Some(inputs) => {
+                        if num_inputs > inputs.len() {
```
I would probably go for option 2: it should be an error if the data in the advice map under `inputs_commitment` doesn't hash to `inputs_commitment`.

bobbinth commented on crates/miden-tx/src/host/mod.rs:
```diff
@@ -509,23 +527,97 @@ where
 
-        Ok(())
+            // Script is present, build the recipient
+            let inputs_data = process.advice_provider().get_mapped_values(&inputs_commitment);
```
Related to one of the above comments: this is the other occurrence of the code that I think could be encapsulated into a helper function.

bobbinth commented on crates/miden-lib/asm/miden/note.masm:
```diff
@@ -91,7 +95,13 @@ end
+#! - Advice map: {
+#!     INPUTS_COMMITMENT: [[NOTE_INPUTS]]
+#!     RECIPIENT: [num_inputs, INPUTS_COMMITMENT, SCRIPT_ROOT, SERIAL_NUM]
+#! }
```
nit: I think this part can go into the inputs/outputs description. For example:

```
#! Inputs:
#!   Operand stack: [inputs_ptr, num_inputs, SERIAL_NUM, SCRIPT_ROOT]
#!   Advice map: {
#!      INPUTS_COMMITMENT: [INPUTS],
#!   }
#! Outputs:
#!   Operand stack: [RECIPIENT]
#!   Advice map: {
#!      INPUTS_COMMITMENT: [INPUTS],
#!      RECIPIENT: [num_inputs, INPUTS_COMMITMENT, SCRIPT_ROOT, SERIAL_NUM],
#!   }
#!
```

Also, I'm not sure `num_inputs` is needed here. I think this has been discussed in another comment, and IIUC, the main reason for this is that sometimes the set of values under `INPUTS_COMMITMENT` is longer than `num_inputs`. But as I mentioned in another comment, I think this behavior is not correct, and I'd rather change it.

Also also, I wonder if we should change the output advice map to follow the approach we use to compute the recipient more closely. Specifically, it is computed as:
```
hash(hash(hash(SERIAL_NUM, [0; 4]), SCRIPT_ROOT), INPUTS_COMMITMENT)
```
So, maybe the entries in the advice map should be:
```
#!   Advice map: {
#!      INPUTS_COMMITMENT: [INPUTS],
#!      RECIPIENT: [SN_SCRIPT_HASH, INPUTS_COMMITMENT],
#!      SN_SCRIPT_HASH: [SN_HASH, SCRIPT_ROOT],
#!      SN_HASH: [SERIAL_NUM, EMPTY_WORD],
#!   }
```

This may be a bit more work, but it feels like cleaner structure where key in every entry is a hash of the corresponding values.

It may also make the code in the procedure below a bit simpler because we'd be able to use the `adv.insert_hdword` instruction and would avoid the need use procedure locals. Specifically, the code could be something like:
```
# => [SERIAL_NUM, SCRIPT_ROOT, INPUTS_COMMITMENT]

padw adv.insert_hdword hmerge
# => [SN_HASH, SCRIPT_ROOT, INPUTS_COMMITMENT]

swapw adv.insert_hdword hmerge
# => [SN_SCRIPT_HASH, INPUTS_COMMITMENT]

swapw adv.insert_hdword hmerge
# => [RECIPIENT]
```

partylikeits1983 replied:
Makes sense! The `build_recipient` procedure is now significantly less complex after implementing your suggestions.

bobbinth commented on crates/miden-objects/src/note/recipient.rs:
```diff
@@ -73,7 +73,7 @@ impl NoteRecipient {
     /// - SCRIPT_ROOT is the commitment of the note script (i.e., the script's MAST root)
     /// - SERIAL_NUMBER is the recipient's serial number
-    pub fn format_for_advice(&self) -> Vec<Felt> {
+    pub fn to_elements(&self) -> Vec<Felt> {
```
If we go through with the suggestion from the previous comment, we won't be able to use this `to_elements()` to reduce the recipient to a sequential list of elements. From what I can tell though, this function is used in only one place `add_output_note_recipient()` and so we could probably move the logic of populating the recipient advice map entries there without too much effort.

bobbinth reviewed (commented):
Looks good! Thank you! I left a few comments inline. The main one is about how we "store" the note recipient in the advice map.

partylikeits1983 commented:
Implemented the suggestions in Bobbin's comment: https://github.com/0xMiden/miden-base/pull/1995#discussion_r2476691328

Restructured the `AdviceMap` layout for `build_recipient` to use a nested structure. This allows the MASM code to use `adv.insert_hdword` and `hmerge` instructions directly, reducing the number of local variables in `build_recipient`.

partylikeits1983 commented:
Refactored `on_note_script_requested` & `NoteRecipient::to_advice_map_entries` to improve logic flow and readability. Also removed excessive comments and duplicate code by extracting helper functions.

partylikeits1983 commented on crates/miden-lib/asm/miden/note.masm:
```diff
@@ -91,7 +95,13 @@ end
+#!
+#! The procedure also inserts the following data into the advice map:
+#! - Advice map: {
+#!     INPUTS_COMMITMENT: [[NOTE_INPUTS]]
```
I think this was a typo. In any case the layout changed, and this was removed.

bobbinth commented on crates/miden-objects/src/transaction/tx_args.rs:
```diff
@@ -165,11 +166,10 @@ impl TransactionArgs {
-            (inputs.commitment(), inputs.format_for_advice()),
-            (script.root(), script_encoded),
-        ];
+        let mut new_elements = note_recipient.to_advice_map_entries();
```
nit: I would maybe move the logic from `to_advice_map_entries()` into this function. It looks simple enough, and I think this is the only place where we use it.

bobbinth commented on crates/miden-tx/src/host/mod.rs:
```diff
@@ -506,23 +524,140 @@ where
+    ///
+    /// If the note script is not present in the advice provider, returns
+    /// [`TransactionEventHandling::Unhandled`] to request the script from the data store.
     fn on_note_after_created(
```
I would re-write this description to make it more clear upfront that we don't always create an `OutputNoteBuilder`. Basically, something alone the lines "If xyz, then we create a new OutputNoteBuilder, otherwise we do abc.".

partylikeits1983 replied:
updated the comment

bobbinth commented on crates/miden-tx/src/host/mod.rs:
```diff
@@ -506,23 +524,140 @@ where
+        process: &ProcessState,
+        inputs_commitment: &Word,
+        num_inputs: usize,
+    ) -> Result<NoteInputs, TransactionKernelError> {
```
I would remove `note_inputs` from the parameters to this method, and instead, would try to determine them here. The way it could work:

1. Try to determine `num_inputs` based on the last non-zero element (same as you have in the code above).
2. Then, build `NoteInputs` and check if the hash matches `inputs_commitment`.
3. If not, increment `num_inputs` by one, and try again.
4. Keep doing this for potentially 6 more times.
5. If `num_inputs` grows to the size of `inputs_data` and there is still no match, return an error.

This is a bit hacky and maybe there are better ways of doing this - but nothing came immediately to mind.

---------

Also nit: I don't think we use `self` in this function - so, it could probably be a "free" function.

bobbinth reviewed (commented):
Looks good! Thank you! I left some comments inline - the main one is about how to read `NoteInputs` from the advice provider.

partylikeits1983 commented on crates/miden-tx/src/host/mod.rs:
```diff
@@ -506,23 +524,140 @@ where
+                .advice_provider()
+                .get_mapped_values(&inputs_commitment)
+                .and_then(|inputs| inputs.iter().rposition(|&x| x != ZERO).map(|pos| pos + 1))
+                .unwrap_or(0);
```
Ah yeah, that is an issue for sure. Implemented this https://github.com/0xMiden/miden-base/pull/1995#discussion_r2479873007 to address this issue.

bobbinth commented on crates/miden-tx/src/host/mod.rs:
```diff
@@ -1129,6 +1213,61 @@ impl<'store, STORE> TransactionBaseHost<'store, STORE> {
+            for attempt in 0..max_attempts {
+                if num_inputs > inputs.len() {
+                    break;
+                }
```
I think we can avoid having `max_attempts` variable here by doing something like:
```rust
while inputs.len() <= num_inputs {
    ...
}
```

bobbinth commented on crates/miden-tx/src/host/mod.rs:
```diff
@@ -930,6 +990,30 @@ where
+    /// # Errors
+    /// Returns an error if the key is not present in the advice map or if the data is malformed
+    /// (not exactly 8 elements).
+    fn read_double_word_from_adv_map(
```
nit: this probably could be a "free" helper function (i.e., it doesn't need to be a method on `TransactionBaseHost`).

bobbinth commented on crates/miden-tx/src/host/mod.rs:
```diff
@@ -1129,6 +1213,61 @@ impl<'store, STORE> TransactionBaseHost<'store, STORE> {
+                if attempt < max_attempts - 1 {
+                    num_inputs += 1;
+                }
+            }
```
I think we can avoid the need for `max_attempts` by rewriting this a bit:
```rust
let mut num_inputs = initial_num_inputs;

loop {
    let candidate_inputs = NoteInputs::new(inputs[0..num_inputs].to_vec())
        .map_err(TransactionKernelError::MalformedNoteInputs)?;

    if candidate_inputs.commitment() == *inputs_commitment {
        return Ok(candidate_inputs);
    }

    num_inputs += 1;
    if num_inputs > inputs.len() {
        break;
    }
}
```

PhilippGackstatter commented on crates/miden-lib/asm/miden/note.masm:
```diff
@@ -114,69 +114,46 @@ end
-    movdn.5 movdn.5
-    # => [RECIPIENT, mem_ptr, end_ptr]
+    padw adv.insert_hdword hmerge
+    # => [SN_HASH, SCRIPT_ROOT, INPUTS_COMMITMENT]
```
Nit: I know the intention is to keep this concise, but could we still extend `SN_HASH` to `SERIAL_HASH` here and everywhere else to make it more readable?

PhilippGackstatter commented on crates/miden-lib/build.rs:
```diff
@@ -162,7 +161,7 @@ fn compile_tx_kernel(source_dir: &Path, target_dir: &Path) -> Result<Assembler>
     kernel_lib.write_to_file(output_file).into_diagnostic()?;
 
-    let assembler = build_assembler(Some(kernel_lib))?;
+    let assembler = build_assembler(Some(kernel_lib))?.with_debug_mode(true);
```
```suggestion
    let assembler = build_assembler(Some(kernel_lib))?;
```

bobbinth commented on crates/miden-objects/src/transaction/tx_args.rs:
```diff
@@ -165,12 +166,19 @@ impl TransactionArgs {
+            (sn_hash, concat_words(note_recipient.serial_num(), Word::empty())),
+            (sn_script_hash, concat_words(sn_hash, script.root())),
+            (note_recipient.digest(), concat_words(sn_script_hash, inputs.commitment())),
         ];
```
My recommendation here was to remove `NoteRecipient::to_advice_map_entries()` as I think the code is simple enough that we probably can handle in this function and also, we can make an argument that how the recipient is converted to advice map entries is concern of the kernel rather than an intrinsic property of the `NoteRecipient`.

bobbinth commented on crates/miden-tx/src/host/mod.rs:
```diff
@@ -1241,6 +1213,61 @@ impl<'store, STORE> TransactionBaseHost<'store, STORE> {
+/// 2. Building NoteInputs and checking if the hash matches inputs_commitment
+/// 3. If not, incrementing num_inputs and trying again (up to 6 more times)
+/// 4. If num_inputs grows to the size of inputs_data and there's still no match, returning an error
+fn extract_note_inputs(
```
This is a good point. I think trial unhashing doesn't give us any benefit here because the input will always be padded to the next multiple of 8 during hashing.

The main issue here is then that we won't be able to determine the "true" number of inputs. For example, the number of inputs may be 5, but if the last input is 0, we'll only pick the first 4. I think this is a broader issue (which we tried to solve a couple of times now) - so, let's create a separate issue to discuss how to fix it (the trivial solution is to just use the RPO padding rule, but that may make things a bit less efficient).

partylikeits1983 replied:
Opened this issue to address this: https://github.com/0xMiden/miden-base/issues/2036

partylikeits1983 commented on crates/miden-tx/src/host/mod.rs:
```diff
@@ -503,26 +521,68 @@ where
+            let script_data = process.advice_provider().get_mapped_values(&script_root);
+
+            if script_data.is_none() {
+                let inputs = extract_note_inputs(process, &inputs_commitment)?;
```
Good idea, refactored remove duplication.
